### ISSUE TITLE
read_parquet is not supported for partitioned parquet

### ISSUE THREAD

**Issue by darrenleeweber** [modin-project/modin#633]
Split from #626 

The `read_parquet` is not supported for a partitioned data set.

### System information

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
$ conda --version
conda 4.6.14
$ python --version
Python 3.7.3
$ pip --version
pip 19.1 from /home/dlweber/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/pip (python 3.7)

$ pip freeze | grep modin
modin==0.5.0
$ pip freeze | grep pandas
pandas==0.24.2
$ pip freeze | grep numpy
numpy==1.16.3
```

miniconda3 was used to install most of the sci-py stack, with a pip clause to add modin, e.g.
```
# environment.yaml
channels:
  - conda-forge
  - defaults

dependencies:
  - python>=3.7
  - affine
  - configobj
  - dask
  - numpy
  - pandas
  - pyarrow
  - rasterio
  - s3fs
  - scikit-learn
  - scipy
  - shapely
  - xarray
  - pip
  - pip:
    - modin
```

### Describe the problem

https://modin.readthedocs.io/en/latest/pandas_supported.html says `read_parquet` is supported, but maybe not for partitioned data.

### error
```
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/backends/pandas/query_compiler.py", line 871, in _full_reduce
    mapped_parts = self.data.map_across_blocks(map_func)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/engines/base/frame/partition_manager.py", line 209, in map_across_blocks
    preprocessed_map_func = self.preprocess_func(map_func)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/engines/base/frame/partition_manager.py", line 100, in preprocess_func
    return self._partition_class.preprocess_func(map_func)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/engines/ray/pandas_on_ray/frame/partition.py", line 108, in preprocess_func
    return ray.put(func)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/ray/worker.py", line 2216, in put
    worker.put_object(object_id, value)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/ray/worker.py", line 375, in put_object
    self.store_and_register(object_id, value)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/ray/worker.py", line 309, in store_and_register
    self.task_driver_id))
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/ray/utils.py", line 475, in _wrapper
    return orig_attr(*args, **kwargs)
  File "pyarrow/_plasma.pyx", line 496, in pyarrow._plasma.PlasmaClient.put
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
  File "pyarrow/serialization.pxi", line 150, in pyarrow.lib.SerializationContext._serialize_callback
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle.py", line 952, in dumps
    cp.dump(obj)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle.py", line 271, in dump
    raise pickle.PicklingError(msg)
_pickle.PicklingError: Could not pickle object as excessively deep recursion required.
```

**Comment by darrenleeweber** [comment]
The following python script works OK for non-partioned data:
```python
import random
import sys

import numpy as np

parquet_path = '/tmp/tmp.parquet'

action = None
if len(sys.argv) > 1:
    action = sys.argv[1]

shape = (10, 10)
size = (1, shape[0] * shape[1])
column_ij = ["%04d_%04d" % (i, j) for i in range(shape[0]) for j in range(shape[1])]

if action == 'save':
    # only import pandas to save, to avoid errors in to_parquet noted in 626
    import pandas as pd
    data = np.random.randint(0, 10000, size=size, dtype="uint16")
    df = pd.DataFrame(data, columns=column_ij)
    df.to_parquet(parquet_path)  # it might need to be partitioned to replicate error?
elif action == 'pandas_read':
    # import pandas to be sure it works without modin
    import pandas as pd
    query_ij = random.sample(column_ij, k=10)
    query_df = pd.read_parquet(parquet_path, columns=query_ij)
    print(query_df)  # print in case it is lazy
elif action == 'modin_read':
    # import modin to override the pandas.read_parquet
    import modin.pandas as pd
    query_ij = random.sample(column_ij, k=10)
    query_df = pd.read_parquet(parquet_path, columns=query_ij)
    print(query_df)  # print in case it is lazy
else:
    raise ValueError("Unknown action, try: save, pandas_read, modin_read")
```

**Comment by darrenleeweber** [comment]
The following partitioned data crashes:

```python
# test_modin.py

import random
import sys

import numpy as np

parquet_path = '/tmp/tmp.parquet'

action = None
if len(sys.argv) > 1:
    action = sys.argv[1]

shape = (10, 10)
size = (1, shape[0] * shape[1])
column_ij = ["%04d_%04d" % (i, j) for i in range(shape[0]) for j in range(shape[1])]

if action == 'save':
    # only import pandas to save, to avoid errors in to_parquet noted in 626
    import pandas as pd
    df_rows = None
    for row in range(10):
        data = np.random.randint(0, 10000, size=size, dtype="uint16")
        data = np.append(data, row)
        data.shape = (1, size[1] + 1)
        df = pd.DataFrame(data, columns=column_ij + ["row"])
        if df_rows is None:
            df_rows = df
        else:
            df_rows = df_rows.append(df)
    df_rows.to_parquet(parquet_path, partition_cols=["row"])  # it might need to be partitioned to replicate error?
elif action == 'pandas_read':
    # import pandas to be sure it works without modin
    import pandas as pd
    query_ij = random.sample(column_ij, k=10)
    query_df = pd.read_parquet(parquet_path, columns=query_ij)
    print(query_df)  # print in case it is lazy
elif action == 'modin_read':
    # import modin to override the pandas.read_parquet
    import modin.pandas as pd
    query_ij = random.sample(column_ij, k=10)
    query_df = pd.read_parquet(parquet_path, columns=query_ij)
    print(query_df)  # print in case it is lazy
else:
    raise ValueError("Unknown action, try: save, pandas_read, modin_read")
```

```text
$ python test_modin.py save

$ tree /tmp/tmp.parquet
/tmp/tmp.parquet
├── row=0
│   └── a1957285c74647f9bb018078e71a647c.parquet
├── row=1
│   └── d11e06018b63405a812bb8dc3fa6a31a.parquet
├── row=2
│   └── 0b1d63ec6b5a4c47bfb7e3fbb9e4a868.parquet
├── row=3
│   └── 7353cb52b6724465a36770d45ef3741b.parquet
├── row=4
│   └── a84039f21d0f4eabbbcd88bb25bcdff2.parquet
├── row=5
│   └── 9b37298627104db9937933beee27ff7e.parquet
├── row=6
│   └── 6d8e0d2cf799462c9029c4b118592f4f.parquet
├── row=7
│   └── 037f6617bc524d32b3221ff476fafea6.parquet
├── row=8
│   └── 33aad5445364496e9e1d0ab709d43bb2.parquet
└── row=9
    └── acb6b6c36aea40b3a8470a5e56f09101.parquet

10 directories, 10 files

$ python test_modin.py pandas_read
   0008_0002  0002_0009  0007_0007  0002_0001  0007_0002  0002_0006  0002_0004  0006_0000  0008_0000  0008_0005 row
0       1456       5945       6271       4104       7659       5349       9462       8449       3591       2538   0
1       2785         20       5421       5529       7902       1859       5324       3323       8436        314   1
2       9065       5116       5318       7186       4043       1089       6023       7076       9760       5045   2
3       1260       4999       1935       1718       6840       5533       2828       7084       8800       7694   3
4       7630       1194        640       5823       9575       3750       1421       3511       2714       7982   4
5       9535       6447        992       1584       7691       9745       5321       6131       2731       5402   5
6       5966       7244       6841        181         53       3490       3345       7568       6808       9004   6
7       1376       3488       3162       8086       2609       2468       5025       8927       5818       2246   7
8       3350       6432        429       9792       8168       4958       8660       4540        646       4235   8
9       2877       4911       9461       3021       2979       7609       1570       3144       1776       9221   9

$ python test_modin.py modin_read
2019-05-22 20:55:49,838	WARNING worker.py:1337 -- WARNING: Not updating worker name since `setproctitle` is not installed. Install this with `pip install setproctitle` (or ray[debug]) to enable monitoring of worker processes.
2019-05-22 20:55:49,838	INFO node.py:469 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-22_20-55-49_12788/logs.
2019-05-22 20:55:49,947	INFO services.py:407 -- Waiting for redis server at 127.0.0.1:45941 to respond...
2019-05-22 20:55:50,060	INFO services.py:407 -- Waiting for redis server at 127.0.0.1:55012 to respond...
2019-05-22 20:55:50,062	INFO services.py:804 -- Starting Redis shard with 2.1 GB max memory.
2019-05-22 20:55:50,068	INFO node.py:483 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-22_20-55-49_12788/logs.
2019-05-22 20:55:50,069	WARNING services.py:1304 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 5238738944 bytes available. This may slow down performance! You may be able to free up space by deleting files in /dev/shm or terminating any running plasma_store_server processes. If you are inside a Docker container, you may need to pass an argument with the flag '--shm-size' to 'docker run'.
2019-05-22 20:55:50,069	INFO services.py:1427 -- Starting the Plasma object store with 6.0 GB memory using /tmp.
Traceback (most recent call last):
  File "test_modin.py", line 41, in <module>
    print(query_df)  # print in case it is lazy
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/pandas/base.py", line 3215, in __str__
    return repr(self)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/pandas/dataframe.py", line 91, in __repr__
    result = repr(self._build_repr_df(num_rows, num_cols))
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/pandas/base.py", line 50, in _build_repr_df
    head_front = head.to_pandas()
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/backends/pandas/query_compiler.py", line 487, in to_pandas
    len(df.index) != len(self.index) or len(df.columns) != len(self.columns)
  File "/home/joe/miniconda3/envs/project/lib/python3.7/site-packages/modin/error_message.py", line 38, in catch_bugs_and_request_email
    "Internal Error. "
Exception: Internal Error. Please email bug_reports@modin.org with the traceback and command that caused this error.
```

**Comment by darrenleeweber** [comment]
If the `columns` selection is dropped, there is a different error:

```text

$ python test_modin.py modin_read
2019-05-22 21:02:36,227	WARNING worker.py:1337 -- WARNING: Not updating worker name since `setproctitle` is not installed. Install this with `pip install setproctitle` (or ray[debug]) to enable monitoring of worker processes.
2019-05-22 21:02:36,228	INFO node.py:469 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-22_21-02-36_12973/logs.
2019-05-22 21:02:36,334	INFO services.py:407 -- Waiting for redis server at 127.0.0.1:36295 to respond...
2019-05-22 21:02:36,445	INFO services.py:407 -- Waiting for redis server at 127.0.0.1:38647 to respond...
2019-05-22 21:02:36,447	INFO services.py:804 -- Starting Redis shard with 2.1 GB max memory.
2019-05-22 21:02:36,453	INFO node.py:483 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2019-05-22_21-02-36_12973/logs.
2019-05-22 21:02:36,454	WARNING services.py:1304 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 5238738944 bytes available. This may slow down performance! You may be able to free up space by deleting files in /dev/shm or terminating any running plasma_store_server processes. If you are inside a Docker container, you may need to pass an argument with the flag '--shm-size' to 'docker run'.
2019-05-22 21:02:36,454	INFO services.py:1427 -- Starting the Plasma object store with 6.0 GB memory using /tmp.
Traceback (most recent call last):
  File "test_modin.py", line 40, in <module>
    query_df = pd.read_parquet(parquet_path)  #, columns=query_ij)
  File "/home/joe/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/modin/pandas/io.py", line 29, in read_parquet
    path=path, columns=columns, engine=engine, **kwargs
  File "/home/joe/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/modin/data_management/factories.py", line 51, in read_parquet
    return cls._determine_engine()._read_parquet(**kwargs)
  File "/home/joe/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/modin/data_management/factories.py", line 55, in _read_parquet
    return cls.io_cls.read_parquet(**kwargs)
  File "/home/joe/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/modin/engines/ray/generic/io.py", line 151, in read_parquet
    pf = ParquetFile(path)
  File "/home/joe/miniconda3/envs/gis-dataprocessing/lib/python3.7/site-packages/ray/pyarrow_files/pyarrow/parquet.py", line 129, in __init__
    self.reader.open(source, use_memory_map=memory_map, metadata=metadata)
  File "pyarrow/_parquet.pyx", line 638, in pyarrow._parquet.ParquetReader.open
  File "pyarrow/io.pxi", line 1271, in pyarrow.lib.get_reader
  File "pyarrow/io.pxi", line 1260, in pyarrow.lib._get_native_file
  File "pyarrow/io.pxi", line 738, in pyarrow.lib.memory_map
  File "pyarrow/io.pxi", line 701, in pyarrow.lib.MemoryMappedFile._open
  File "pyarrow/error.pxi", line 83, in pyarrow.lib.check_status
pyarrow.lib.ArrowIOError: Memory mapping file failed: No such device

```

**Comment by devin-petersohn** [comment]
Thanks @darrenleeweber, is this the same issue as #624? That is currently being worked on by @williamma12

**Comment by williamma12** [comment]
Hi @darrenleeweber, I was able to reproduce the issue and tried it with #632. #632 fixes the issue if `columns` selection is turned off. I will look into fixing it when there is `column` selection later as well

**Comment by williamma12** [comment]
@darrenleeweber I'm not too sure where the pickling error is coming from but #632 should fix both errors that you were running into

**Comment by devin-petersohn** [comment]
Resolved by #632
